src/bosh-director/lib/bosh/director/api/task_remover.rb

```diff
@@ -29,14 +47,39 @@ def remove_task(task)
 
     private
 
-    def removal_candidates_dataset(type)
+    def removal_max_tasks_candidates_dataset(type)
       base_filter = Bosh::Director::Models::Task.where(type: type)
         .exclude(state: %w[processing queued])
         .select(:id, :output).order { Sequel.desc(:id) }
-
       starting_id = base_filter.limit(1, @max_tasks).first&.id || 0
 
       base_filter.where { id <= starting_id }
     end
+
+    def removal_retention_candidates_dataset(type)
+      retention_time = DateTime.now.to_time - convert_to_time_duration(@retention_period)
+      Bosh::Director::Models::Task.where(type: type)
+                                  .where { checkpoint_time < retention_time }
+                                  .exclude(state: %w[processing queued])
+                                  .select(:id, :output)
+    end
+
+    def removal_deployment_retention_candidates_dataset(type, deployment_with_retention_period)
+      retention_time = DateTime.now.to_time - convert_to_time_duration(deployment_with_retention_period['retention_period'])
+      Bosh::Director::Models::Task.where(type: type)
+                                  .where(deployment_name: deployment_with_retention_period['deployment'])
+                                  .where { checkpoint_time < retention_time }
+                                  .exclude(state: %w[processing queued])
+                                  .select(:id, :output)
+    end
+
+    def convert_to_time_duration(string)
+      multipliers = { "d" => 24*60*60, "h" => 60*60, "m" => 60, "s" => 1 }
+      segments = string.scan(/(\d+)([a-z])/)
+
+      segments.inject(0) do |total, (value, unit)|
+        total + value.to_i * multipliers[unit]
```

Comment:
I see 2 issues here:
- we are not validating if the configuration given satisfy the units accessed in the hash: 'd', 'h', 'm' and 's'
- Is it really needed to provide such a detailed retention time? Why don't we just fix the value to be in days?

Reply:
Good point. I have chosen days as default unit!